src/main/java/org/broadinstitute/hellbender/utils/iterators/ShardingIterator.java

```diff
@@ -0,0 +1,132 @@
+    private final SAMSequenceDictionary dictionary;
+
+    /**
+     * Wraps an iterator to perform sharding.
+     *
+     * @param iterator   sorted iterator to iterate over windows.
+     * @param boundaries sorted list of shard boundaries to iterate over.
+     * @param dictionary dictionary for position comparison.
+     */
+    public ShardingIterator(final Iterator<T> iterator,
+            final List<ShardBoundary> boundaries,
+            final SAMSequenceDictionary dictionary) {
+        // check arguments
+        Utils.validateArg(Utils.nonNull(iterator, "null iterator").hasNext(), "empty iterator");
+        Utils.nonEmpty(boundaries, "Empty shard boundaries");
+        Utils.nonNull(dictionary, "null dictionary");
+
+        // TODO: avoid new object creation for already PeekableIterator and LinkedList args
+        // store the iterator and allow peek()
+        this.it = new PeekableIterator<>(iterator);
+        // TODO: avoid new object creation for already LinkedList and LinkedList args
+        // convert the boundaries into a queue
+        this.shards = new LinkedList<>(boundaries);
+        // store the dictionary
+        this.dictionary = dictionary;
+        // create a new queue for the next data
+        this.nextData = new LinkedList<>();
+
+        // start the iteration
+        advance();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return !shards.isEmpty();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Note: for overlapping shards, the objects stored in two (or more) consecutive shards are
+     * references and not clones. If modified in the previous shard, the state would be kept in the
+     * next returned value. It is up to the caller to make copies if necessary to avoid state in the
+     * next iteration.
+     */
+    @Override
+    public Shard<T> next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+        // copying the deque to being able to remove unnecesary data and fill in with the next
+        final Shard<T> next = new ShardBoundaryShard<T>(shards.poll(),
+                nextData.isEmpty() ? Collections.emptyList() : new LinkedList<>(nextData));
+        advance();
+        return next;
+    }
+
+    // advance to the next data
+    private void advance() {
+        // if we already finished the iteration, we clean up the data
+        if (shards.isEmpty()) {
+            // TODO: should we also close the iterator?
+            nextData.clear();
+        } else {
+            // first we empty the queue if it has non-overlapping data with the next shard
+            removeNonOverlapping();
+            // then, fill in the data
+            fillQueueWithOverlapping();
+        }
+    }
+
+    // assumes that shards are not empty
+    private void removeNonOverlapping() {
+        while (!nextData.isEmpty() && !nextData.peek()
+                .overlaps(shards.peek().getPaddedInterval())) {
+            nextData.poll();
```

Comment:
This is not quite right. The reads could be of varying lengths (because of clipping, for example), and so you can't stop as soon as you encounter the *first* read that overlaps the next shard's interval, because there could be a *shorter* read after it that doesn't overlap the next interval.

You need to check *all* reads in `nextData` that start before the next shard's padded interval, and remove those that don't overlap with it. You can stop only once you encounter a read whose start position is >= the start of the next shard's padded interval. See `FeatureCache.trimToNewStartPosition()` for a similar algorithm.

Once you've made this change, you should add a unit test that covers the edge case described above.

Reply:
Good catch! I am working now on this, and I will let you know when it is prepared for next review.

Reply:
Done (failing test in one commit, and fix in the next)